### PR TITLE
Identity hash code substitution

### PR DESCRIPTION
**Pull request by zhelenskiy**
Substitute identity hashcode, received on the first run, on the following runs.

### CODE REVIEW

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/traceDebugger/HashCodeTests.kt [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+package org.jetbrains.kotlinx.lincheck_test.traceDebugger
+
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import java.util.*
+
+abstract class HashCodeTest : AbstractNativeCallTest()
+
+class SimpleHashCodeTest : HashCodeTest() {
+    @Operation
+    fun operation() = List(100) { Any().hashCode() }
```
[the remark this answers is not in the document]
Comment (zhelenskiy) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/traceDebugger/HashCodeTests.kt:20]: Yes

[the remark this answers is not in the document]
Comment (eupp) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/traceDebugger/HashCodeTests.kt:20]: I am sorry, but I just don't see where it is checked in the code. Could you please clarify this?

[the remark this answers is not in the document]
Comment (zhelenskiy) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/traceDebugger/HashCodeTests.kt:20]: See `AbstractNativeCallTest`.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/traceDebugger/AbstractNativeCallTest.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                .verifier(FailingVerifier::class.java)
+                .customize()
+                .checkImpl(this::class.java) { lincheckFailure ->
+                    val results = lincheckFailure?.results?.parallelResults?.flatten()?.takeIf { it.isNotEmpty() }
+                    require(results != null) { lincheckFailure.toString() }
+                    if (shouldFail()) {
+                        require(results.all { it is ExceptionResult })// { lincheckFailure.toString() }
+                    } else {
+                        require(results.none { it is ExceptionResult })// { lincheckFailure.toString() }
+                    }
```
[the remark this answers is not in the document]
Comment (zhelenskiy) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/traceDebugger/AbstractNativeCallTest.kt:53]: Done

File: bootstrap/src/sun/nio/ch/lincheck/Injections.java [hunk 3]
```diff
@@ -293,6 +293,14 @@ public static void beforeNewObjectCreation(String className) {
     public static void afterNewObjectCreation(Object obj) {
         getEventTracker().afterNewObjectCreation(obj);
     }
+    
+    public static long getNextObjectId() {
```
[the remark this answers is not in the document]
Comment (zhelenskiy) [reply bootstrap/src/sun/nio/ch/lincheck/Injections.java:300]: Done

File: bootstrap/src/sun/nio/ch/lincheck/Injections.java [hunk 4]
```diff
@@ -293,6 +293,14 @@ public static void beforeNewObjectCreation(String className) {
...
+        return getEventTracker().getNextObjectId();
+    }
+    
+    public static void advanceCurrentObjectIdWithKnownOldObjectId(long oldId) {
```
[the remark this answers is not in the document]
Comment (zhelenskiy) [reply bootstrap/src/sun/nio/ch/lincheck/Injections.java:311]: Done

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectInitialHashCodes.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+     * Resets ids numeration and starts them from 0.
+     */
+    fun resetObjectIds() {
+        if (isLoggingEnabled()) {
+            println("Restart counter")
+        }
+        nextObjectId.set(0)
+    }
+    
+    fun advanceCurrentObjectIdWithKnownOldObjectId(oldObjectId: Id) {
```
[the remark this answers is not in the document]
Comment (zhelenskiy) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectInitialHashCodes.kt:71]: Done

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectInitialHashCodes.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            println(currentObjectId)
+            println(initialIdentityHashCode)
+        }
+        // TODO: bizarre and crazy code below (might not work for all JVM implementations)
+        UnsafeHolder.UNSAFE.putInt(obj, IDENTITY_HASHCODE_OFFSET, initialIdentityHashCode)
+
+        return currentObjectId
+    }
+
+    private fun isLoggingEnabled(): Boolean = System.getProperty("lincheck.log", "false") == "true"
```
[the remark this answers is not in the document]
Comment (zhelenskiy) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectInitialHashCodes.kt]: Done

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectInitialHashCodes.kt [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+ */
+
+package org.jetbrains.kotlinx.lincheck.strategy.managed
+
+import org.jetbrains.kotlinx.lincheck.util.UnsafeHolder
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicLong
+
+
+private typealias Id = Long
```
[the remark this answers is not in the document]
Comment (zhelenskiy) [reply src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectInitialHashCodes.kt:18]: Done

**Comment by ndkoval** [comment]
@zhelenskiy, could you please make a more intuitive title for the PR and add a description?

**Comment by zhelenskiy** [comment]
@ndkoval done

**Pull request merged**